feat: precondition guides (#91)
sapphiredev/website: docs/Guide/preconditions/reporting-precondition-failure.mdx

---
sidebar_position: 2
title: Reporting precondition failure
---

When a precondition fails, it's usually important for the user to know why. For example, if they hit a cooldown or lack
permissions, that should somehow be communicated. However, by default, nothing will happen if a precondition blocks a
message.

To change this, we'll need to create a `commandDenied` listener, which is triggered when a precondition fails. For more
information on how to create listeners, see the [`Creating Listeners`][listeners] section.

:::caution

The `commandDenied` event shouldn't be confused with the `commandError` event, which is triggered when a command throws
an error.

:::

`commandDenied` supplies the following information: the [`UserError`][error] that was created from the precondition, and
the [`CommandDeniedPayload`][payload], which includes necessary context.

> ```typescript ts2esm2cjs

Reviewer: ```suggestion
```typescript ts2esm2cjs|{4}|{5}
```

import type { UserError, CommandDeniedPayload } from '@sapphire/framework';
import { Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {
  public run(error: UserError, { message }: CommandDeniedPayload) {
    // ...
  }
}
```

The `message` property of the `error` parameter will include the error message, as the name suggests. In [Creating
Preconditions][creating-preconditions], you can find that we defined this property within the `this.error()` method!

There are many possibilities for what you can do with the error, but the simplest is to just send it directly to the
user. That is what we'll do in this example:

> ```typescript ts2esm2cjs|{6}|{6}

Reviewer: ```suggestion
```typescript ts2esm2cjs|{5}|{5}
```

<Tabs groupId="language-choice">
<TabItem value="javascript" label="JavaScript" default>

```javascript {5}
const { Listener } = require('@sapphire/framework');

module.exports = class CommandDeniedListener extends Listener {
  run(error, { message }) {
    if (error.context.silent) return;
    return message.channel.send(error.message);
  }
};
```

</TabItem>
<TabItem value="ESM" label="ESM">

```javascript {5}
import { Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {
  run(error, { message }) {
    if (error.context.silent) return;
    return message.channel.send(error.message);
  }
}
```

</TabItem>
<TabItem value="typescript" label="TypeScript">

```typescript {8}
import { UserError, CommandDeniedPayload, Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {
  public run(error: UserError, { message }: CommandDeniedPayload) {
    // A bit of a hack is required for TypeScript
    // Since `error.context` is of type `unknown`
    if (Reflect.get(Object(error.context), 'silent')) return;
    return message.channel.send(error.message);
  }
}
```

</TabItem>

> </Tabs>

Reviewer: Tabs left here, replace with `ts2esm2cjs`

Author: The problem is that the typescript code isn't the same as the javascript

Reviewer: Uh.... hmm.. well Reflect.get works just fine in JS as well so lets do it like so:

---

```ts

import { UserError, CommandDeniedPayload, Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {
  public run(error: UserError, { message }: CommandDeniedPayload) {
    if (Reflect.get(Object(error.context), 'silent')) return;
    return message.channel.send(error.message);
  }
}
```

:::note

In the code block above we use `if (Reflect.get(Object(error.context), 'silent'))` as opposed to `if (error.context.silent)` for TypeScript. When writing JavaScript code you can use the latter just fine.

To clarify this, with TypeScript `error.context` has the type `unknown` therefore trying to do `error.context.silent` will make TypeScript error on trying to read property `silent` of type `unknown`.

:::

    public run(error: UserError, { message }: CommandDeniedPayload) {
        // ...
    }
}
```

</TabItem>
</Tabs>

The `message` property of the `error` parameter will include the error message, as the name suggests.
In [Creating Preconditions][creating-preconditions], you can find that we defined this property
within the `this.error()` method!

There are many possibilities for what you can do with the error, but the simplest is to just send it directly to the user.
That is what we'll do in this example:

<Tabs groupId="language-choice">
<TabItem value="javascript" label="JavaScript" default>

```javascript {5}
const { Listener } = require('@sapphire/framework');

module.exports = class CommandDeniedListener extends Listener {
    run(error, { message }) {
        return message.channel.send(error.message);
    }
};
```

</TabItem>
<TabItem value="ESM" label="ESM">

```javascript {5}
import { Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {
    run(error, { message }) {
        return message.channel.send(error.message);
    }
}
```

</TabItem>
<TabItem value="typescript" label="TypeScript">

> ```typescript {5}

Reviewer 2: ```suggestion
```typescript {4}
```

import type { UserError, CommandDeniedPayload } from '@sapphire/framework';
import { Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {
    public run(error: UserError, { message }: CommandDeniedPayload) {
        return message.channel.send(error.message);
    }
}
```

</TabItem>
</Tabs>

## Ignoring Precondition Failures

> If someone who isn't a bot owner tries to use a command intended only for the bot owner, sometimes you don't want to

Reviewer 2: ```suggestion
If someone who isn't a bot owner tries to use a command which is intended only for the bot owner, sometimes you don't want to
```

send a message notifying them that they don't have permission. Instead, you'd rather let the command be blocked silently.
To do this, we can make use of the [`context`][context] property of `UserError`s. This property aims to contain information
about the context in which the error was thrown, and the value can be absolutely anything.

> We can take advantage of this by passing the value `{ silent: true }` to the [`this.error()`][preconditionError] call of

Reviewer 2: ```suggestion
We can take advantage of this by passing `{ silent: true }` to the [`this.error()`][preconditionError] call of
```

> We'll use the `OwnerOnly` precondition we made in[Creating Preconditions][creating-preconditions] to demonstrate this.

Reviewer 3: ```suggestion
We'll use the `OwnerOnly` precondition we made in [Creating Preconditions][creating-preconditions] to demonstrate this.
```

module.exports = class OwnerOnlyPrecondition extends Precondition {
  public run(message) {
    return message.author.id === 'YOUR_ID'
      ? this.ok()
      : this.error({ 
          message: 'Only the bot owner can use this command!' 
          context: { silent: true }
        })
  }
};
```

</TabItem>
<TabItem value="ESM" label="ESM">

```javascript {9}
import { Precondition } from '@sapphire/framework';

export class OwnerOnlyPrecondition extends Precondition {
  public run(message) {
    return message.author.id === 'YOUR_ID'
      ? this.ok()
      : this.error({ 
          message: 'Only the bot owner can use this command!',
          context: { silent: true }
        })
  }
}
```

</TabItem>
<TabItem value="typescript" label="TypeScript">

```typescript {10}
import type { Message } from '@sapphire/framework';
import { Precondition } from '@sapphire/framework';

export class OwnerOnlyPrecondition extends Precondition {
  public run(message: Message) {
    return message.author.id === 'YOUR_ID'
      ? this.ok()
      : this.error({ 
          message: 'Only the bot owner can use this command!'
          context: { silent: true }
        })
  }
}
```
</TabItem>
</Tabs>

> We can then detect this property in our listener, and ignore the failure if we find it.

Reviewer 2: ```suggestion
We can then check if this property exists on the error in our listener, and ignore the failure if we find it.
```

<Tabs groupId="language-choice">
<TabItem value="javascript" label="JavaScript" default>

```javascript {5}
const { Listener } = require('@sapphire/framework');

module.exports = class CommandDeniedListener extends Listener {
    run(error, { message }) {
        if (error.context.silent) return;
        return message.channel.send(error.message);
    }
};
```

</TabItem>
<TabItem value="ESM" label="ESM">

```javascript {5}
import { Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {

>     run(error, { message }) {

Reviewer 2: useless destructuring?

Author: `message.channel.send` lol

        if (error.context.silent) return;
        return message.channel.send(error.message);
    }
}
```

</TabItem>
<TabItem value="typescript" label="TypeScript">

```typescript {8}
import type { UserError, CommandDeniedPayload } from '@sapphire/framework';
import { Listener } from '@sapphire/framework';

export class CommandDeniedListener extends Listener {

>     public run(error: UserError, { message }: CommandDeniedPayload) {

Reviewer 2: context is undefined
```suggestion
    public run({ message , context}: CommandDeniedPayload) {
```

Author: Sorry, it should've been `error.context`. (`context` isn't actually in `CommandDeniedPayload`, nor would it be the first parameter)

Reviewer 2: Then fix it however you want and resolve this conflix also error,  { dest} won't work because error is a single argument not two, am I right @favna

Reviewer 2: https://github.com/Spectera-bot/Spectera/blob/main/src/listeners/commands/commandDenied.ts#L4

    // A bit of a hack is required for TypeScript

>     // Since `error.context` is of type `unknown`

Reviewer 3: We could always make CommandDeniedPayload take in a generic for the context, which would be better than this hack

Author: I'll submit a pr hehe

title: Reporting precondition failure
---

<<<<<<< HEAD

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

`commandDenied` supplies the following information: the [`UserError`][error] that was created from the precondition, and
the [`CommandDeniedPayload`][payload], which includes necessary context.

<<<<<<< HEAD
```typescript ts2esm2cjs

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

<TabItem value="typescript" label="TypeScript">

```typescript

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

}
```

<<<<<<< HEAD

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

</TabItem>
</Tabs>

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

The `message` property of the `error` parameter will include the error message, as the name suggests. In [Creating
Preconditions][creating-preconditions], you can find that we defined this property within the `this.error()` method!

There are many possibilities for what you can do with the error, but the simplest is to just send it directly to the
user. That is what we'll do in this example:

<<<<<<< HEAD
```typescript ts2esm2cjs|{6}|{6}
import { Listener, UserError, CommandDeniedPayload } from '@sapphire/framework';

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

```typescript {6}
import type { UserError, CommandDeniedPayload } from '@sapphire/framework';
import { Listener } from '@sapphire/framework';

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

}
```

<<<<<<< HEAD

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

</TabItem>
</Tabs>

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

We can take advantage of this by adding `context: { silent: true }` to the [`this.error()`][preconditionerror] options.
We'll use the `OwnerOnly` precondition we made in [Creating Preconditions][creating-preconditions] to demonstrate this.

<<<<<<< HEAD
```typescript ts2esm2cjs|{9}|{9}

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

<TabItem value="typescript" label="TypeScript">

```typescript {9}

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

}
```

<<<<<<< HEAD

> =======

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)

</TabItem>
</Tabs>

> >>>>>>> bce70940bf8edd1967c694d4ec08344839c08ecd

Reviewer: ![](https://play.pokemonshowdown.com/sprites/ani/ditto.gif)
